### DISCUSSION TITLE
Reset-Password api -generate a post req for generate link for update password to keyclaok  having 403 error

### DISCUSSION THREAD

**Discussion by ParthaPratima** [Sunbird-Lern/Community#111] [Query] [answered]
POST  :  http://localhost:8080/auth/realms/<realm-name>/get-required-action-link

{
    "redirectUri": "https://dev.sunbirded.org/resources",
    "clientId": "<client-id name>",
    "expirationInSecs": "155520000",
    "userName": "<user-name>",
    "requiredAction": "UPDATE_PASSWORD"
}

ERROR:: 
{
    "errorMessage": "Does not have realm admin role."
}
Could someone please hep me in resolving this issue.We are using the existing sunbird code.

**Comment by reshmi-nair** [comment, marked as answer]
@ParthaPratima Please check whether below fields are configured in userorg service environment.

sunbird_sso_url: url for keycloak server
sunbird_sso_realm: keycloak realm name
sunbird_ssorname: keycloak _username: keycloak user name
sunbird_sso_password: keycloak password
sunbird_sso_client_id: key cloak client id
sunbird_sso_client_secret : keycloak client secret

Sample pwd reset api call to userorg service after otp verfication.
/user/v1/password/reset api
{
"request": {
"type": "email",
"key": "",
"userId": "",
"reqData": "{"iv":"4c6b802dd9255f79917cad9d4f2b0ed7a","encryptedData":"1427443f81af4edb37cf5614501c20fb1670bdae8791ccbfdabc17558e260fc6d744c2f94b53c29a3bed46f795e1c802a51e3560e80b6c1b3c88d568e1d94f281bbe4d4bddeeed7f35461be467ecd3ec9adb2529e866688407139f4bd8faf6c01"}"
}
}
This api response will give the keycloak password reset url.
sample url:
/auth/realms/sunbird/login-actions/required-action?execution=UPDATE_PASSWORD&client_id=&tab_id=E5GtSHoRxTk

So make sure you have set the above environment variables, so that you can get the proper reset pwd link.

**Reply by ParthaPratima** [reply]
the above variables are already added to the envinonment , inspite of that it show 403 , 
{
error : doesn't have realm admin role
}
THe decoded Token ::
{
  "jti": "0f82f573-5c21-4ccd-bee8-0373d2dc8953",
  "exp": 1682682422,
  "nbf": 0,
  "iat": 1682678822,
  "iss": "http://localhost:8080/auth/realms/sunbird",
  "aud": "realm-management",
  "sub": "55d4c2da-e31f-4d6f-ba56-51ee1aa4a3bd",
  "typ": "Bearer",
  "azp": "sunbird",
  "auth_time": 0,
  "session_state": "ad4bab22-7ea6-4ab1-a69c-109abab6300b",
  "acr": "1",
  "realm_access": {
    "roles": [
      "offline_access"
    ]
  },
  "resource_access": {
    "realm-management": {
      "roles": [
        "view-identity-providers",
        "view-realm",
        "manage-identity-providers",
        "impersonation",
        "realm-admin",
        "create-client",
        "manage-users",
        "query-realms",
        "view-authorization",
        "query-clients",
        "query-users",
        "manage-events",
        "manage-realm",
        "view-events",
        "view-users",
        "view-clients",
        "manage-authorization",
        "manage-clients",
        "query-groups"
      ]
    },
    "sunbird": {
      "roles": [
        "realm-admin",
        "admin"
      ]
    }
  },
  "scope": "profile email",
  "email_verified": true,
  "preferred_username": "sunbirduser2"
}
but after decode the access Token "realm-admin" role is present inside the Token ,...please provide any other approaches to resolve this issue

**Reply by ParthaPratima** [reply]
or any conf changes that it going to be worked

**Reply by ParthaPratima** [reply]
inside reset-password api it generate a link for update password which shows 403 {doesn't have realm admin role }

**Reply by reshmi-nair** [reply]
@ParthaPratima Can we connect and solve this? because if the configurations are correct the code should work fine. 
Reach out to me at reshmi.nair@tarento.com
